File: test/ui.test.js
const { test, expect } = require('@playwright/test');
const pageUrl = 'http://localhost:8000';

test('Verify "All Books" is visible', async ({ page }) => {
    // Increase the timeout if needed
    await page.goto(pageUrl, { waitUntil: 'domcontentloaded' });
    await page.waitForSelector('nav.navbar');
    const allBooksLink = await page.$('a[href="/catalog"]');
    const isAllBookLinkVisible = await allBooksLink.isVisible();
    expect(isAllBookLinkVisible).toBe(true);
});

test('Verrifi Login button is visible', async ({ page }) => {
    await page.goto(pageUrl, { waitUntil: 'domcontentloaded' });
    await page.waitForSelector('nav.navbar');
    const loginButton = await page.$('a[href="/login"]');
    const isLoginButtonVisible = await loginButton.isVisible();
    expect(isLoginButtonVisible).toBe(true);})

test('Verrifi Register button is visible', async ({ page }) => {
    await page.goto(pageUrl, { waitUntil: 'domcontentloaded' });
    await page.waitForSelector('nav.navbar');
    const registerButton = await page.$('a[href="/register"]');
    const isRegisterButtonVisible = await registerButton.isVisible();
    expect(isRegisterButtonVisible).toBe(true);})

test('Verrifi "All Books" link is visible after user login', async ({ page }) => {
    await page.goto('http://localhost:8000/login', { waitUntil: 'domcontentloaded' }); 

    await page.fill('input[name="email"]', 'peter@abv.bg');
    await page.fill('input[name="password"]', '123456');
    await page.click('input[type="submit"]');
    const allBooksLink = await page.$('a[href="/catalog"]');
    const isAllBookLinkVisible = await allBooksLink.isVisible();
    expect(isAllBookLinkVisible).toBe(true);})

test('Login with valid credentials', async ({ page }) => {
    await page.goto('http://localhost:8000/login', { waitUntil: 'domcontentloaded' }); 

    await page.fill('input[name="email"]', 'peter@abv.bg');
    await page.fill('input[name="password"]', '123456');
    await page.click('input[type="submit"]');
    await page.$('a[href="/catalog"]');
    expect(page.url()).toBe('http://localhost:8000/catalog');})

test('Login with empty credentials', async ({ page }) => {
    await page.goto('http://localhost:8000/login', { waitUntil: 'domcontentloaded' }); 
    await page.click('input[type="submit"]');

    page.on('dialog', async dialog => {
        expect(dialog.type()).toContain('alert');
        expect(dialog.message()).toContain('All fields are required');
        await dialog.accept();
    });
        await page.$('a[href="/login"]');
        expect(page.url()).toBe('http://localhost:8000/login');
});

test('Register with empty credentials', async ({ page }) => {
    await page.goto('http://localhost:8000/register', { waitUntil: 'domcontentloaded' }); 
    await page.click('input[type="submit"]');

    page.on('dialog', async dialog => {
        expect(dialog.type()).toContain('alert');
        expect(dialog.message()).toContain('All fields are required');
        await dialog.accept();
    });
        await page.$('a[href="/register"]');
        expect(page.url()).toBe('http://localhost:8000/register');
});

test('Add book with correct data', async ({ page }) => {
    await page.goto('http://localhost:8000/login', { waitUntil: 'domcontentloaded' }); 

    await page.fill('input[name="email"]', 'peter@abv.bg');
    await page.fill('input[name="password"]', '123456');
    await Promise.all([
        page.click('input[type="submit"]'),
        page.waitForURL('http://localhost:8000/catalog')
    ])
    await page.click('a[href="/create"]');
    await page.waitForSelector('#create-form');
    await page.fill("#title", 'Test Book');
    await page.fill("#description", 'This is a test book Description');
    await page.fill("#image", 'http://example.com/book-image.jpg');
    await page.selectOption('#type', 'Fiction');
    await page.click('#create-form input[type="submit"]');
    await page.waitForURL('http://localhost:8000/catalog');
    expect(page.url()).toBe('http://localhost:8000/catalog');
})

test('Add book with empty title field', async ({ page }) => {
    await page.goto('http://localhost:8000/login', { waitUntil: 'domcontentloaded' }); 

    await page.fill('input[name="email"]', 'peter@abv.bg');
    await page.fill('input[name="password"]', '123456');
    await Promise.all([
        page.click('input[type="submit"]'),
        page.waitForURL('http://localhost:8000/catalog')
    ])
    await page.click('a[href="/create"]');
    await page.waitForSelector('#create-form');
    // await page.fill("#title", 'Test Book');
    await page.fill("#description", 'This is a test book Description');
    await page.fill("#image", 'http://example.com/book-image.jpg');
    await page.selectOption('#type', 'Fiction');

    page.on('dialog', async dialog => {
        expect(dialog.type()).toContain('alert');
        expect(dialog.message()).toContain('All fields are required');
        await dialog.accept();
    });

    await page.$('a[href="/create"]');
    
    expect(page.url()).toBe('http://localhost:8000/create');
});

test('Login and verify all books are displayed', async ({ page }) => {
    await page.goto('http://localhost:8000/login', { waitUntil: 'domcontentloaded' }); 

    await page.fill('input[name="email"]', 'peter@abv.bg');
    await page.fill('input[name="password"]', '123456');
    await Promise.all([
        page.click('input[type="submit"]'),
        page.waitForURL('http://localhost:8000/catalog')
    ])

    await page.waitForSelector('.dashboard');
    const bookElements = await page.$$('.other-books-list li');
    expect(bookElements.length).toBeGreaterThan(0);

});

test('Login and navigate to Details page', async ({ page }) => {
    await page.goto('http://localhost:8000/login', { waitUntil: 'domcontentloaded' }); 

    await page.fill('input[name="email"]', 'peter@abv.bg');
    await page.fill('input[name="password"]', '123456');
    await Promise.all([
        page.click('input[type="submit"]'),
        page.waitForURL('http://localhost:8000/catalog')
    ])
    await page.click('a[href="/catalog"]');
    await page.waitForSelector('.otherBooks');
    await page.click('.otherBooks a.button');
    await page.waitForSelector('.book-information');
    const detailsPageTitle = await page.textContent('.book-information h3');
    expect(detailsPageTitle).toBe('Test Book');
});

test('Verify redirection of Logout link after user is logged in', async ({ page }) => {
    await page.goto('http://localhost:8000/login', { waitUntil: 'domcontentloaded' }); 

    await page.fill('input[name="email"]', 'peter@abv.bg');
    await page.fill('input[name="password"]', '123456');
    await page.click('input[type="submit"]');

    const logoutLink = await page.$('a[href="javascript:void(0)"]');
    await logoutLink.click();

    const redirectedUrl = page.url();
    expect(redirectedUrl).toBe('http://localhost:8000/');

})

File: test/ui.test.py
import re
from playwright.sync_api import Page, expect

def test_has_title(page: Page):
    page.goto("http://localhost:8000/")

    # Expect a title "to contain" a substring.
    expect(page).get_by_text(re.compile("All Books Login Register Welcome"))

def test_login_button(page: Page):
    page.goto("http://localhost:8000/")

    # Click the get started link.
    page.get_by_role("link", name="Login").click()

    # Expects page to have a heading with the name of Installation.
    expect(page.get_by_placeholder("Email", name="Email")).to_be_visible()





    # page.goto("http://localhost:8000/")
    # page.get_by_text("All Books Login Register Welcome, {email} My Books Add Book Logout").click()
    # page.get_by_role("link", name="Login").click()
    # page.get_by_role("link", name="Register").click()
    # page.get_by_role("link", name="All Books").click()
    # page.locator("li").filter(has_text="To Kill a Mockingbird Type: Classic Details").get_by_role("link").click()
    # page.get_by_text("The unforgettable novel of a childhood in a sleepy Southern town and the crisis ").click()
    # page.get_by_role("heading", name="Description:").click()
    # page.get_by_text("Type: Classic").click()
    # page.get_by_role("heading", name="To Kill a Mockingbird").click()
    # page.get_by_text("Likes: 0").click()
    # page.get_by_role("img").nth(1).click()
    # page.get_by_role("link", name="Login").click()
    # page.get_by_placeholder("Email").click()
    # page.get_by_placeholder("Email").fill("peter@abv.bg")
    # page.get_by_placeholder("Password").click()
    # page.get_by_placeholder("Password").fill("123456")
    # page.get_by_role("button", name="Login").click()
    # page.get_by_role("link", name="My Books").click()
    # page.get_by_role("link", name="Add Book").click()
    # page.get_by_text("Title").click()
    # page.get_by_text("Description").click()
    # page.get_by_text("Image").click()
    # page.get_by_text("Type").click()
    # page.get_by_text("Welcome, peter@abv.bg").click()
    # page.get_by_role("link", name="Logout").click()
    # page.get_by_role("link", name="All Books").click()
    # page.get_by_role("link", name="Login").click()
    # page.get_by_role("button", name="Login").click(button="right")
    # page.get_by_role("link", name="Register").click()
    # page.once("dialog", lambda dialog: dialog.dismiss())
    # page.get_by_role("button", name="Register").click()

    # ---------------------
   
# import re
# from playwright.sync_api import Page, expect

# def test_has_title(page: Page):
#     page.goto("https://playwright.dev/")

#     # Expect a title "to contain" a substring.
#     expect(page).to_have_title(re.compile("Playwright"))

# def test_get_started_link(page: Page):
#     page.goto("https://playwright.dev/")

#     # Click the get started link.
#     page.get_by_role("link", name="Get started").click()

#     # Expects page to have a heading with the name of Installation.
#     expect(page.get_by_role("heading", name="Installation")).to_be_visible()